server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/MDNSendMethod.scala

```diff
@@ -212,21 +212,17 @@ class MDNSendMethod @Inject()(serializer: MDNSerializer,
       .map(mailbox => mailbox.getAddress)
       .toRight(MDNSendNotFoundException("Invalid \"Disposition-Notification-To\" header field."))
 
-  private def getMDNFinalRecipient(requestEntry: MDNSendCreateRequest, identity: Identity): Either[MDNSendForbiddenFromException, FinalRecipient] = {
-    if (requestEntry.finalRecipient.isEmpty) {
-      scala.Right(FinalRecipient.builder()
-        .finalRecipient(Text.fromRawText(identity.email.asString()))
-        .build())
-    }
-    else {
-      val tryMailAddress: Try[MailAddress] = requestEntry.finalRecipient.get.getMailAddress
-      if (tryMailAddress.isSuccess && tryMailAddress.get.equals(identity.email)) {
-        scala.Right(requestEntry.finalRecipient.get.asJava.get)
-      } else {
-        Left(MDNSendForbiddenFromException("The user is not allowed to use the given \"finalRecipient\" property"))
+  private def getMDNFinalRecipient(requestEntry: MDNSendCreateRequest, identity: Identity): Either[MDNSendForbiddenFromException, FinalRecipient] =
+    requestEntry.finalRecipient
+      .map(finalRecipient => finalRecipient.getMailAddress)
+      .map(mayBeMailAddress => (mayBeMailAddress.isSuccess && mayBeMailAddress.get.equals(identity.email)))
```

Comment:
We should be able to do some kind more advanced map here;

```
.map {
   case scala.Right(finalRecipient) if finalRecipient.equals(identity.email) => scala.Right(finalRecipient.asJava.get)
   case scala.Right(finalRecipient) => Left(MDNSendForbiddenFromException("The user is not allowed to use the given \"finalRecipient\" property"))
   case Left(e) => Left(e)
}
    .getOrElse(scala.Right(FinalRecipient.builder()
        .finalRecipient(Text.fromRawText(identity.email.asString()))
        .build()))
```

Yes scala is super expressive!